### ISSUE TITLE
False positive in undocumented_unsafe_blocks (stable-to-stable regression)

### ISSUE THREAD

**Issue by shssoichiro** [rust-lang/rust-clippy#9142]
### Summary

When defining a variable using an unsafe block, if the unsafe block is not on the same line as the variable definition, the undocumented_unsafe_blocks triggers a warning when enabled. This did not trigger the lint in Rust 1.61. It does trigger the lint in Rust 1.62 as well as the latest nightly.

### Lint Name

undocumented_unsafe_blocks

### Reproducer

I tried this code:

```rust
// SAFETY: We write to the array below before reading from it.
let mut col_coeffs_backing: Aligned<[i32; 64]> =
  unsafe { Aligned::uninitialized() };
```

I saw this happen:

```
warning: unsafe block missing a safety comment
   --> src/transform/forward.rs:129:9
    |
129 |         unsafe { Aligned::uninitialized() };
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
note: the lint level is defined here
   --> src/lib.rs:69:9
    |
69  | #![warn(clippy::undocumented_unsafe_blocks)]
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    = help: consider adding a safety comment on the preceding line
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#undocumented_unsafe_blocks
```

I expected to see no warning.

In Rust 1.61, there is no warning, as expected. Additionally, keeping the code on one line, or moving the SAFETY comment down one line, both resolve the warning. e.g. the following two examples both produce no warning:

```rust
// SAFETY: We write to the array below before reading from it.
let mut col_coeffs_backing: Aligned<[i32; 64]> = unsafe { Aligned::uninitialized() };
```

```rust
let mut col_coeffs_backing: Aligned<[i32; 64]> =
  // SAFETY: We write to the array below before reading from it.
  unsafe { Aligned::uninitialized() };
```

### Version

```text
Reproducible on both:

rustc 1.62.0 (a8314ef7d 2022-06-27)
binary: rustc
commit-hash: a8314ef7d0ec7b75c336af2c9857bfaf43002bfc
commit-date: 2022-06-27
host: x86_64-unknown-linux-gnu
release: 1.62.0
LLVM version: 14.0.5


and:

rustc 1.64.0-nightly (06754d885 2022-07-08)
binary: rustc
commit-hash: 06754d8852bea286a3a76d373ccd17e66afb5a8b
commit-date: 2022-07-08
host: x86_64-unknown-linux-gnu
release: 1.64.0-nightly
LLVM version: 14.0.6
```

### Additional Labels

@rustbot +label regression-from-stable-to-nightly regression-from-stable-to-stable

**Comment by anall** [comment]
Regressed in 65f96e2b53369f6b403b655b2212272c2ae13d42

**Comment by smmalis37** [comment]
Just hit this too, has any progress been made?

**Comment by denfren** [comment]
FYI @llogiq  - It seems I have found another case that also doesn't seem to work even when the definition is one line. However it looks like your fix covers this as well, right?

```
use std::ffi::CStr;

// SAFETY: This is safe.
const NAME: &CStr = unsafe {CStr::from_bytes_with_nul_unchecked(b"Hello World\0") };
```
tested on [Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=40e06da3585b47d086d55e83b8d6c862)
 with `clippy 0.1.66`
